File: app/javascript/controllers/order_items_controller.js
import { Controller } from "@hotwired/stimulus";

// Connects to data-controller="order-items"
export default class extends Controller {
  static targets = ["productVariantId", "unitPrice", "quantity", "subtotal"];

  async connect() {
    await this.fetchOrderItem();
  }

  async change() {
    await this.fetchOrderItem();
  }

  async fetchOrderItem() {
    const productVariantId = this.productVariantIdTarget.value;
    if (!productVariantId) return;

    const queryString = new URLSearchParams({
      product_variant_id: productVariantId,
      quantity: this.quantityTarget.value,
    }).toString();

    try {
      const response = await fetch(`/order_items/new?${queryString}`, {
        method: "GET",
        headers: { Accept: "application/json" },
      });

      if (!response.ok) {
        throw new Error(`Response status: ${response.status}`);
      }

      const orderItem = await response.json();
      const { unit_price: unitPrice, subtotal } = orderItem;

      this.unitPriceTarget.value = unitPrice;
      this.subtotalTarget.value = subtotal;
    } catch (error) {
      console.error(error.message);
    }
  }
}

File: app/controllers/order_items_controller.rb
class OrderItemsController < ApplicationController
  def new
    @order_item = OrderItem.new(order_item_params)
    @order_item.unit_price = product_variant.price.to_d
    @order_item.subtotal = product_variant.price.to_d * @order_item.quantity.to_i

    render json: @order_item
  end

  private

  def order_item_params
    params.permit(:product_variant_id, :quantity)
  end

  def product_variant
    @product_variant ||=
      ProductVariantRepo.new(ProductVariant)
        .by_company(current_company)
        .find_or_initialize_by(id: params[:product_variant_id])
  end
end
